src/test/hosting/runTags.spec.ts

```diff
@@ -0,0 +1,307 @@
+          return { regex: ".*", run: r };
+        }),
+      },
+      createTime: "now",
+      createUser: {
+        email: "inlined@gmail.com",
+      },
+      fileCount: 0,
+      versionBytes: 0,
+    };
+  }
+
+  function service(id: string, ...tags: Array<string | runNS.TrafficTarget>): runNS.Service {
+    return {
+      apiVersion: "serving.knative.dev/v1",
+      kind: "Service",
+      metadata: {
+        name: id,
+        namespace: PROJECT,
+        labels: {
+          [runNS.LOCATION_LABEL]: REGION,
+        },
+      },
+      spec: {
+        template: {
+          metadata: {
+            name: "revision",
+            namespace: "project",
+          },
+          spec: {
+            containers: [],
+          },
+        },
+        traffic: [
+          {
+            latestRevision: true,
+            percent: 100,
+          },
+          ...tags.map((tag) => {
+            if (typeof tag === "string") {
+              return {
+                revisionName: `revision-${tag}`,
+                tag: tag,
+                percent: 0,
+              };
+            } else {
+              return tag;
+            }
+          }),
+        ],
+      },
+      status: {
+        observedGeneration: 50,
+        latestCreatedRevisionName: "latest",
+        latestRevisionName: "latest",
+        traffic: [
+          {
+            revisionName: "latest",
+            latestRevision: true,
+            percent: 100,
+          },
+          ...tags.map((tag) => {
+            if (typeof tag === "string") {
+              return {
+                revisionName: `revision-${tag}`,
+                tag: tag,
+                percent: 0,
+              };
+            } else {
+              return {
+                percent: 0,
+                ...tag,
+              };
+            }
+          }),
+        ],
+        conditions: [],
+        url: "https://google.com",
+        address: {
+          url: "https://google.com",
+        },
+      },
+    };
+  }
+
+  beforeEach(() => {
+    // We need the library to attempt to do something for us to observe side effects.
+    run = sinon.stub(runNS);
+    hosting = sinon.stub(hostingNS);
+    runTags = sinon.stub(runTagsNS);
+
+    hosting.listSites.withArgs(PROJECT).resolves([site]);
+    hosting.listVersions.rejects(new Error("Unexpected hosting.listSites"));
+
+    run.getService.rejects(new Error("Unexpected run.getService"));
+    run.updateService.rejects(new Error("Unexpected run.updateService"));
+    run.gcpIds.restore();
+
+    runTags.ensureLatestRevisionTagged.throws(
+      new Error("Unexpected runTags.ensureLatestRevisionTagged")
+    );
+    runTags.gcTagsForServices.rejects(new Error("Unepxected runTags.gcTagsForServices"));
+    runTags.setRewriteTags.rejects(new Error("Unexpected runTags.setRewriteTags call"));
+    runTags.setGarbageCollectionThreshold.restore();
+  });
+
+  afterEach(() => {
+    sinon.verifyAndRestore();
```

Comment:
I don't see that these tests use mocks (`sinon.mock`), so `sinon.restore()` is sufficient.